doc/plutus-core-spec/builtins-future.tex

```diff
@@ -0,0 +1,413 @@
+\newcounter{notenumberC}
+\renewcommand{\note}[1]{
+  \bigskip
+  \refstepcounter{notenumberC}
+  \noindent\textbf{Note \thenotenumberC. #1}
+}
+
+\newpage
+
+\section{Future Built-in Types and Functions}
+\label{appendix:default-builtins-future}
+A future release will add support for BLS12-381 elliptic curve pairing
+operations
+(see~\cite{BLS12-381}, \cite[4.2.1]{IETF-pairing-friendly-curves}, \cite{BLST-library}).
+This involves adding three new types and seventeen new built-in functions.  The
+description of the semantics of these types and functions is quite complex and
+requires a considerable amount of notation, most of which is used only in this
+section.
+
+\subsection{Built-in types}
+\label{sec:future-built-in-types}
+Tables~\ref{table:future-built-in-types} describes three new built-in
+types.
+
+\newcommand{\MlResult}{H}
+\newcommand{\Fq}{\mathbb{F}_q}
+\newcommand{\Fqq}{\mathbb{F}_{q^2}}
+\newcommand{\FF}{\mathbb{F}_{q^{12}}}
+
+\begin{table}[H]
+  \centering
+    \begin{tabular}{|l|p{2cm}|l|}
+        \hline
+        Type & Denotation & Concrete Syntax\\
+        \hline
+        $\ty{bls12\_381\_G1\_element}$ &   $G_1$ & \texttt{0x[0-9A-Fa-f]\{96\}} \text{(see Note~\ref{note:bls-syntax})}\\
+        $\ty{bls12\_381\_G2\_element}$ &   $G_2$ & \texttt{0x[0-9A-Fa-f]\{192\}} \text{(see Note~\ref{note:bls-syntax})}\\
+        $\ty{bls12\_381\_MlResult}$    &   $\MlResult$  &  None (see Note~\ref{note:bls-syntax})\\
+        \hline
+    \end{tabular}
+    \caption{Atomic Types}
+    \label{table:future-built-in-types}
+\end{table}
+
+
+%% \paragraph{$G_1$ and $G_2$}.
+\noindent Here $G_1$ and  $G_2$ are both cyclic groups of prime order $r$, where 
+$$
+r = \mathtt{0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001}.
+$$
+        
+\noindent To define these groups we need the finite field $\Fq$ where
+\begin{align*}
+q = \mathtt{0x}&\mathtt{1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf}\\
+              &\mathtt{6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab}
+\end{align*}
+
+\noindent which is a 381-bit prime. The field $\Fq$ is isomorphic to $\Z_q$,
+the ring of integers modulo $q$, and hence there is a natural epimorphism from
+$\Z$ to $\Fq$ which we denote by $n \mapsto \bar{n}$.  Given $x \in \Fq$, we
+denote by $\tilde{x}$ the smallest non-negative integer $n$ with $\bar{n} = x$.
+We sometimes use literal integers to represent elements of $\Fq$ in the obvious
+way.
+
+We also make use of the field $\Fqq = \Fq[X]/(X^2+1)$; we may regard $\Fqq$ as
+the set $\{a+bu: a,b \in \Fq\}$ where $u^2=-1$.
+
+It is convenient to say that an element $a$ of $\Fq$ is \textit{larger} than
+another element $b$ (and $b$ is \textit{smaller} than $a$) if $\tilde{a}
+> \tilde{b}$ in $\Z$.  We extend this terminology to $\Fqq$ by saying that
+$a+bu$ is larger than $c+du$ if either $b$ is larger than $d$ or $b=d$ and $a$
+is larger than $c$.  \red{Is this the right way round? Ie, we compare the
```

Comment:
We compare the `u` parts first. See the [zkcash repo](https://github.com/zkcrypto/bls12_381/blob/main/src/fp2.rs#L178): 

```rust
self.c1.lexicographically_largest()
            | (self.c1.is_zero() & self.c0.lexicographically_largest())
```

and c1 is the `u` part as seen [here](https://github.com/zkcrypto/bls12_381/blob/main/src/fp2.rs#L18).

Reply:
> We compare the u parts first,
Thanks.  I wanted to be sure I'd got that right.